File: gdtracker/src/util/taskStatus.ts
import type { TaskStatus } from '../api/tasks'

export const allStatus: TaskStatus[] = ['PENDING', 'TODO', 'IN_PROGRESS', 'COMPLETED', 'DONE']

export function statusLabel(s: TaskStatus): string {
    switch (s) {
        case 'PENDING':
            return 'Pending'
        case 'TODO':
            return 'Todo'
        case 'IN_PROGRESS':
            return 'In Progress'
        case 'COMPLETED':
            return 'Completed'
        case 'DONE':
            return 'Done'
        default:
            return s
    }
}

export function nextTaskStatus(s: TaskStatus): TaskStatus | null {
    const i = allStatus.indexOf(s)
    if (i < 0 || i >= allStatus.length - 1) return null
    return allStatus[i + 1]!
}

File: gdtracker-go-api/internal/httpserver/taskstatus.go
package httpserver

// allowedTaskStatuses matches gdtracker-api TaskStatus enum.
var allowedTaskStatuses = map[string]struct{}{
	"PENDING":     {},
	"TODO":        {},
	"IN_PROGRESS": {},
	"COMPLETED":   {},
	"DONE":        {},
}

// orderedTaskStatuses is workflow order (lowest to highest).
var orderedTaskStatuses = []string{
	"PENDING",
	"TODO",
	"IN_PROGRESS",
	"COMPLETED",
	"DONE",
}

func isValidTaskStatus(s string) bool {
	_, ok := allowedTaskStatuses[s]
	return ok
}

func taskStatusRank(s string) (int, bool) {
	for i, v := range orderedTaskStatuses {
		if v == s {
			return i, true
		}
	}
	return 0, false
}

// minTaskStatus returns the lowest workflow status among statuses, or "" if none are valid.
func minTaskStatus(statuses []string) string {
	best := -1
	for _, s := range statuses {
		r, ok := taskStatusRank(s)
		if !ok {
			continue
		}
		if best < 0 || r < best {
			best = r
		}
	}
	if best < 0 {
		return ""
	}
	return orderedTaskStatuses[best]
}
